### DISCUSSION TITLE
Combining the Log and Script Window

### DISCUSSION THREAD

**Discussion by rdstern** [IDEMSInternational/R-Instat#9830] [General]
@ChrisMarsh82 and @lloyddewit and @N-thony and @rachelkg Please can we discuss and plan and implement some changes for the March release?  We are, at last making progress on the help system, and videos.  So I would like plans soon, and carewfully, even if implementation may take considerably longer.

![image](https://user-images.githubusercontent.com/11226469/221354652-04fbf803-60be-40ee-91b4-21b7a4388992.png)

The current system of Windows is as follows:

**NOTE THAT THE ACTIONABLE POINTS TO THIS ARE IN ISSUE #9831**

- [ ] a) **_the data window_** does not have a direct Help button.
- [x] b) It has similar right-click menus to Excel, so at the top, (for operations on variables), left, (for operations on rows) and bottom (for operations on data frames)  
- [x] c) Each right-click menu has a Help option.  This goes to a right-click option in the help file.
- [ ] d) The **_column metadata_** window has 2 of the same right-click menus ( column and data frame) as the data window.  The help from there goes to the metadata window instead.
- [x] e) The **_data frame metadata_** window has the same data frame, right-click menu as the data window and the column metadata window, but without the help option.
- [x] f) Right-clicking within the **_column metadata or data frame metada_** window, just gives a single Help item (shown in the figure above, and goes to the corresponding window in the help.
- [x] g) The new script windows have a help button that goes to the script window in the help.
- [x] h) It also has a right-click menu, with a help button and that goes to the script window too.
- [x] i) The log window has a right-click menu with a help option that goes to the correct window in the help.  (I hope we can also consider whether this window can now be combined with the script windows, or it should remain separate?)
- [x] j) The output window has buttons at the top, but no help option.  
- [ ] K) It does not have a right-click menu.
- [x] l) The help menu has a windows option

**Comment by rdstern** [comment]
My initial very incomplete suggestions are that:
a) We move the current Right-Click Menus from within the Data View section to below the Help menu section. 
b) Within that menu we go to a separate entry for each different right-click menu.  So there is a Column, Row, Data Frame, Script, etc item for each one.
c) etc???

**Comment by lloyddewit** [comment]
> I hope we can also consider whether this window can now be combined with the script windows, or it should remain separate?

@rdstern For me this makes sense. Do you think David or Danny has a strong opinion?
@ChrisMarsh82 @N-thony @rachelkg If you have a view, then please speak now :)

**Comment by lloyddewit** [comment]
@rdstern I spoke briefly with @dannyparsons and @ChrisMarsh82 this morning. Both were happy with making the log window a tab of the script window. @dannyparsons requested that the log tab has a grey background (and/or other visual clues) to signify that it is read-only and slightly different to the other tabs. He also requested that the button and menu option used to open the log window should remain the same.
I added a note to issue #7989 .

**Comment by lloyddewit** [comment]
@rdstern If we combine the log/script windows, then we have 5 windows - 3 grid windows ('data', 'column metadata' and 'data frame meta data'), and 2 non-grid windows ('output' and 'log/script').

These 5 windows should have consistent help functionality:
- Each window should have a help button. The output and log/script windows should have a button at the top-right. The grid windows should have a button at the bottom right (there doesn't seem to be space at the top).
- The help button would open the help page for that window. This help page would include all the help information for the window, including a description of any right mouse button menu options.
- If each window, has a help button, then having a help option in the right mouse button menu just clutters up the interface and is potentially confusing to the user. I suggest that we remove all the right mouse button help menu options. The only exception could be for the grid windows. If we decide that there is no space for a help button at the bottom of the grid, then we could have a right mouse button help menu option instead (but only for the grid windows). In general, I think we should minimize the right mouse button menu options because these options are normally not visible to the user.

What do you think?

**Comment by rdstern** [comment]
@lloyddewit on your points above
@dannyparsons said he had discussed combining the log and script windows.  We discussed the details and suggest as follows:
a) On the View menu there is now a single item, namely `Log/Script Window ` 
b) On the toolbar there is the same as before.  I assume pressing the icon itself will go to the current tab?
c) But there could still be the pull-down which would now say Log Tab and Script Tab. It would again return to the one that was last open.  Perhaps initially it goes to the log, because there isn't anything in any script? (I suppose if it is easy, and there are multiple script tabs, with names, then maybe it could permit any script tab to be chosen.  Presumably up to a limit - like the recent data frames?)
d) the tooltip changes to say Log/Script Window.

Now to the windows.  This is timely because I weant to also discuss the R-Instat windows in the help file and also have a script.

a) There are 7 windows, namely our 5 - as you say - and 2 from R.  I suggest we keep the 2 R windows, at least for now.  
b) The R data window is on the bottom right-click and also on Prepare >Data Frame > View.  It is just a viewer, and we don't use it so much now that Chris has made all the improvements to the grid.  But it is still useful to be able to see all the data together.
c) The R-viewer is again used less now that Patrick has his Maximi9se window.  But it is still accessible and has options, like resizing graphs and printing them in different formats, that are not yet available in Patrick's new window.  
d) Agree with you on a help button for each window, and - for all except the grid right-click - there is no help option on any  right-click menu.  The windows help can say all that is needed about the right-click menus, and they are obvious items.
e) I suggest the right-click menus related to the grid are sufficiently complicated, that the Help is sensible there.  There are 3 right-click menus and they "go over" the 3 data windows.

**Comment by rdstern** [comment]
@lloyddewit I continue the windows discussion, because I can't count!  There are actually 9 windows in R-Instat!  In your 5 you have omitted Patrick's new Maximise window - which still needs improvement.  I explain below as a first draft of text, and also a possible video.  I hope it is clear and very simple for you - otherwise I need to explain it very differently.

**_The R-Instat windows_**

Here is R-Instat with data from the two tutorials together:

![image](https://user-images.githubusercontent.com/11226469/221804431-a7f2e44b-71de-40bc-9803-99e24c38ce1b.png)

This shows the 2 main windows.  On the left is the main data window and the output, or results are in the window on the right.

The data window has tabs at the bottom, to move from one data set to another.  For larger data sets it is just a window onto part of the data, aand you can use the arrows at the bottom to move from one page to another.

The output, or results window also has tabs.  All output goes to the main tab, but you can copy important results to additional tabs as you wish.

You usually use dialogues with R-Instat and each dialogue gives one way to access two further windows.

![image](https://user-images.githubusercontent.com/11226469/221807408-2f5e9265-1c3f-4dd6-8c3e-fed27d6f544b.png)

At the bottom is a Help button.  Pressing that button opens the `Help window` and takes you straight to the Help on that dialogue.

![image](https://user-images.githubusercontent.com/11226469/221808081-cf98dfad-a0e1-430d-8e03-45e90260d78a.png)

An alternative route to this window is the Help menu.

![image](https://user-images.githubusercontent.com/11226469/221809100-670822a7-cc0e-41ba-8c94-4f1eb63ddc7a.png)

This allows you to choose your entry point into the Help window.

Returning to the dialogue the bottom right button is for more advanced users and is always labelled `To Script`.  It is enabled (as is the Ok button) whenever the dialogue is complete.  Pressing this button opens the log/script window as follows:

![image](https://user-images.githubusercontent.com/11226469/221814639-1f44ee27-8a75-47c2-8f47-ff78d368b2d5.png)

This window also has tabs.  One is for the log, which automatically keeps a record of the R commands that have corresponded to your dialogue use of R-Instat.  Then you can have one, or more, tabs to examine, edit and try different R scripts.

The View menu provides another route to the Log/Script window as follows:

![image](https://user-images.githubusercontent.com/11226469/221816349-a7279021-5929-4f4e-b46a-8f910f36ba61.png)

These are the main windows, but there are also 3 more windows to complete your picture of the data, and two more for the outputs.  

The data first.  Start by returning to the default layout - that's via the curly arrow on the toolbar.  Or, from the View menu, press to return to the default layout.

Then, possibly from the toolbar click on the i icon to get more information about the data.  This opens the Column metadata window. as follows:

This is another tabbed window that corresponds to the tabs on the data window.  For a given data frame, this window has one row for each variable.  So, the diamonds data, with 10 variables, has the same 10 rows in this window.  It gives the details of each variable.  One property of the variables, that isn't used much in R, but we like, is the variable labels.  It is often convenient to have reasonably short variable names plus a description.  If you ever import survey data from SPSS, or other software, then 

Add another set to show this - from sjmisc say

Then, either from the toolbar, or from the View menu, add the data frame metadata as well.  This has one row for each data frame, and includes details of their dimensions.

Finally, on the data side, R provides a viewer, if you would like to see all your data together.  This is eaither via a right-click on the bottom tab - with the View Data Frame option.  Or it is also from the Prepare > Data Frame > View Data dialogue.

On the Output, or results, side there is a Maximise button associated with each output.  This opens a secondary output window  - what do we call it?  This is for any results, either numeric or graphical.  It is particularly useful for pivot/tables, charts, etc.

I need to describe this now.  It also provides a home for plotly and enables these outputs, particularly pivot tables to be saved as html.  You can have lots of them open, but they each appear as saeparate windows - should they be tabbed!!!

An alternative, for graphs, is the R graph viewer.  This is resizable - also dangerous - and permits graphs to be saved in various formats.  I hope we can add to our maximised window, so we don't need the R graphics window?  Is that a reasonable action.

**Comment by lloyddewit** [comment]
@rdstern I think that all the requests above related to the Log/script window have been implemented. So I removed the `log_script_window` label (otherwise this issue is listed when I search for open Log/script related issues).

**Comment by ARMSTRONGOPONDO** [comment]
@rdstern
@Ag-Derek
@lilyclements

The current updated system of Windows is as follows:
a) The data window still doesn't have a direct Help button.
b) It has similar right-click menus  at the top (for operations on variables), left (for operations on rows), 
and different right-click menus at the bottom (for operations on data frames)

c) Each right-click menu has a Help option. This goes to a right-click option in the help file in the data window

d) The column metadata window has 2 of the same right-click menus (rows and data frame) as the data window, but I have to click the far end on the left along the rows for the menu to appear. The help on clicking it also goes to R packages and commands. The column variable is not clickable in the column metadata

![Image](https://github.com/user-attachments/assets/f5b2fbf2-a114-4e2d-9fd7-3ba4ea207e73)
![Image](https://github.com/user-attachments/assets/ca71eae1-1c99-46b3-8c42-a4fc35ffc2c9)

e)The data frame metadata window has the same data frame, right-click menu as the data window, and the column metadata window. It also has the help menu, but when you click it, it takes you to the  R packages and commands under Gridextra

![Image](https://github.com/user-attachments/assets/d9025a16-86ae-454f-b28a-b337723f193b)

![Image](https://github.com/user-attachments/assets/ca71eae1-1c99-46b3-8c42-a4fc35ffc2c9)


f) Right-clicking within the column metadata or data frame metadata window just gives a single Help item (shown in the figure below)
When I click it, it takes me to R packages and commands under Gridextra

![Image](https://github.com/user-attachments/assets/ce018f9c-b966-4478-9ef7-578c8a01a2e4)
![Image](https://github.com/user-attachments/assets/ca71eae1-1c99-46b3-8c42-a4fc35ffc2c9)

g) The new script windows have a help button that goes to the script window in the help.
h) It also has a right-click menu, with a help button, and that goes to the script window too.
i) The log window has a right-click menu with a help option that goes to the correct window in the help.

 (I hope we can also consider whether this window can now be combined with the script windows, or it should remain separate?)
This is not implemented.

j) The output window has buttons at the top, with the help option.
K) It does not have a right-click menu.
l) The help menu has a Windows option

**Comment by lilyclements** [comment]
@ARMSTRONGOPONDO this is a great review of what is remaining here. I am going to move your remaining actionable points to a new issue. Then I will move this into a discussion to continue the discussions on if the log window and script windows should be combined or separate.


Not complete are points:
**a** (see b/c for an example of it working)
**d** (and you've added further bits to be completed)
**k**
